<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
		<head>
				<link href="style/css/default.css" rel="stylesheet" type="text/css">

		<!-- Required scripts -->
				<script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
				<script type="text/javascript" src="js/jquery.ba-hashchange.min.js"></script>
				<script type="text/javascript" src="js/local.js"></script>

		<!-- MathJax configuration-->
		<script type="text/x-mathjax-config">
			MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
		</script>				 <!-- Google Analytics tracking code-->
		<script type="text/javascript"
			src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
		</script>

		<!-- Google analytics tracking code-->
				<script type="text/javascript">
						var _gaq = _gaq || [];
						_gaq.push(['_setAccount', 'UA-16521993-1']);
						_gaq.push(['_trackPageview']);

						(function() {
								var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
								ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
								var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
						})();
				</script>

		<!-- Go to first page if we get a vanilla URL-->
				<script type="text/javascript">
						$(document).ready(
								function() {
										setFirstPage( "overview" ) ;
								}
						) ;
				</script>
				<title></title>
		</head>
		<body>
				<div class="header" id="header1">
						<div class="header_text">
				bingwa
			</div>
						<div class="header_subtext">
				model-based meta-analysis for genome-wide association studies
						</div>
				</div>
				<div class="button_bar">
						<div class="nav_button" name="overview">
								<span class = "nav_button_content" >
										overview
								</span>
						</div>
						<div class="nav_button" name="documentation">
								<span class = "nav_button_content" >
										documentation
								</span>
						</div>
						<div class="nav_button" name="tutorial">
								<span class = "nav_button_content" >
										tutorial
								</span>
						</div>
						<div class="nav_button" name="file_formats">
								<span class = "nav_button_content" >
										file formats
								</span>
						</div>
						<div class="nav_button" name="download">
								<span class = "nav_button_content" >
										download
								</span>
						</div>
				</div>
				<div class="main_text">
						<div id="overview" style="display:none">
								<p>
					<em>This page documents version 0.9-dev of bingwa, which is currently considered experimental</em>.
				</p>
								<p>
					<strong>Introduction.</strong>
					<span class="program_name">bingwa</span> is a command-line tool used to carry out meta-analysis of genome-wide association studies.
					It is particularly suited for use with the program <a href="https://mathgen.stats.ox.ac.uk/genetics_software/snptest/snptest.html">SNPTEST</a>,
					but can also be used with other association testing programs such as <a href="http://www.helsinki.fi/~mjxpirin/download.html">MMM</a>.
					Features of bingwa include:
				</p>
				<ul>
					<li>
						Direct support for SNPTEST output files.
					</li>
					<li>
						Support for fixed-effect frequentist and a flexible bayesian meta-analysis method.
					</li>
					<li>
						Support for both univariate and multivariate meta-analysis (including support
						for meta-analysing tests performed using	<a href="https://mathgen.stats.ox.ac.uk/genetics_software/snptest/snptest.html#multinomial_tests">SNPTEST's multinomial test</a>).
					</li>
				</ul>
								<p>
										<strong>Disclaimer.</strong>
					This page documents version 0.9-dev of bingwa which is currently considered <b>experimental</b>.
					This means we expect some features not to work, or
					not to work well, or to work wrongly, or to destroy your computer or sanity.
								</p>
								<p>
					<strong>References.</strong>
					If you make use of bingwa, please cite the following paper, which describes the key methodology underlying the program:
					<ul>
						<li>
							"<em>A novel locus of resistance to severe malaria in a region of ancient balancing selection</em>",
							Malaria Genomic Epidemiology Network, Nature (2015)
						</li>
				</ul>
				</p>
				<p>
					The core ideas underlying bingwa were originally developed and applied in these papers:
				<ul>
					<li>
						"<em>Genome-wide association study identifies a variant in HDAC9 associated with large vessel ischemic stroke</em>",
						International Stroke Genetics Consortium (ISGC) and Wellcome Trust Case Control Consortium 2 (WTCCC2), Nat. Genetics (2012)
					</li>
					<li>"<em>Reappraisal of known malaria resistance loci in a large multi-centre study</em>", Rockett et al, Nat. Genetics (2014).</li>
					<li>
						"<em>Imputation-Based Meta-Analysis of Severe Malaria in Three African Populations</em>",
						Band et al, PLOS Genetics (2013)
					</li>
				</ul>
								<p>
										<strong>Acknowledgements.</strong> Bingwa was developed at the <a href="http://www.well.ox.ac.uk/home">Wellcome Trust Centre for Human Genetics</a>, Oxford, UK.
					A full list of people who contributed to the methodology will appear here.
								</p>
						</div>
						<div id="documentation" style="display:none">
								<div class="display_diagram_right">
										<!--<object data="style/images/inthinnerator-flow.svg"	type="image/svg+xml" align="center">
					</object>-->
								</div>
								<p>(This page is under construction.)</p>
								<p>
										The general process followed by bingwa is as follows.	 Briefly, bingwa proceeds by
					<ol>
						<li>Loading association test data from a set of cohorts (e.g. SNPTEST files).</li>
						<li>Matching variants between cohorts.</li>
						<li>Computing frequentist meta-analysis and one or more Bayes factors.</li>
						<li>Writing results.</li>
					</ol>
								</p>
				<p>
					We'll update this page with more details on the method soon.	In the meantime the <a href="#tutorial">tutorial</a>
					has usage examples.
								</p>
			</div>
			<div id="tutorial" style="display:none">
				<h2>
					Running bingwa
				</h2>
								<div class="tutorial_header">
										This page shows common command lines to use bingwa.	 These examples assume the program is being run from a directory
					containing two SNPTEST files, <span class="monospace">cohort1.snptest</span> and <span class="monospace">cohort2.snptest</span>.
					You can see example files in the <a href="example/">examples directory</a>.
								</div>
								<div class="tutorial_separator">
								</div>
								<div class="task">
										<div class="task_name">
												View the program usage page
										</div>
										<div class="task_command_line">
												$ bingwa -help | less
										</div>
								</div>
								<div class="task">
										<div class="task_name">
												Compute fixed-effect meta-analysis of the two files
										</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite
											</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest cohort2.snptest -o output.csv
										</div>
										<div class="task_notes">
												<b>Note</b>: bingwa determines the format of the output file by looking at the file extension.
												The first command above writes output to a sqlite3 file (more details on this below), while the
												second command writes to a csv file.	Also supported are space-separated (.txt) or tab-separated (.tsv) files.
												The format <code>-o sqlite://&lt;filename&gt;</code> can be used to force sqlite3 output format.
										</div>
								</div>
								<div class="task">
										<div class="task_name">
												Compute a fixed-effect bayesian meta-analysis
										</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -prior fixed:sd=0.2,0.2/cor=1,1,1 
					</div>
										<div class="task_notes">
												<b>Note</b>: The above specifies a multivariate normal (MVN) prior on the 'true' effects
						that is centred at zero and has variance-covariance matrix with every entry equal 0.2<sup>2</sup>.
						(Bingwa will print this model to the screen output so you can double-check this).
						Under this model the true effects in the two populations are assumed perfectly correlated; the only
						variation comes from estimation effects represented by the within-study standard errors.
						The first part of the prior specification (before the colon) is a user-specifiable name for the model.
						In this case the bayes factor for the model will be stored in a column named <code>fixed:bf</code>.
										</div>
								</div>
								<div class="task">
										<div class="task_name">
												Compute both a fixed and independent-effect bayesian meta-analysis
										</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -prior fixed:sd=0.2,0.2/cor=1,1,1 -prior independent:sd=0.2,0.2/cor=1,0,1 
					</div>
										<div class="task_notes">
												<b>Note</b>: In the above command the
						output will contain <code>fixed:bf</code> and <code>independent:bf</code> columns.
						Here, the independent model assumes that true effects are uncorrelated between the cohorts - i.e.
						they have diagonal variance-covariance matrix.
						In addition to the two model-specific bayes factors, bingwa will compute an equally weighted model-averaged bayes factor
						over all the models specified, named <code>mean_bf</code>.
										</div>
								</div>
								<div class="task">
										<div class="task_name">
												Using model shorthand
										</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest ... cohort&lt;n&gt;.snptest -o output.sqlite -prior fixed:tau=1/sd=0.2/cor=1 -prior independent:tau=0/sd=0.2/cor=1
					</div>
										<div class="task_notes">
											<p>
												<b>Note</b>: In the above command we use a short-hand version of the prior specification,
												identified because it is of the form <code>tau=...</code>.
												Here, the first part (tau) denotes the between-cohort correlation, while the rest now
												denotes the prior covariance matrix within each cohort.	 In this case there is one 
												effect parameter in each cohort, so the prior covariance is the 1x1 matrix (0.2^2),
												and the correlation between cohorts is 1 (for fixed-effect model) or 0 (for independent-effect model).
											</p>
											<p>
												While the difference in this example are small, this feature comes into its own when
												many cohorts are analysed.	However, we always advise always checking the screen output to debug model definitions.
											</p>
										</div>
									</div>
								<div class="task">
									<div class="task_name">
											Loading priors from a file
										</div>
										<div class="task_command_line">
											$ bingwa -data cohort1.snptest ... cohort&lt;n&gt;.snptest -o output.sqlite -priors priors.txt
										</div>
										<div class="task_notes">
											<p>
												The <code>-priors</code> option tells bingwa to load prior specifications from a file.
												Priors in the file should be specified as on the command line, but with a couple of additional
												features:
											</p>
											<ul>
												<li>Any line starting with # is ignored - this allows models to be commented.</li>
												<li>Line breaks (newlines) within model specifications are ignored if they
													follow one of the 'punctuation' characters - i.e. immediately after a <code>:</code>, <code>/</code>, or comma character.
													This allows long models to be formatted across lines.
												</li>
											</ul>
											<p>
												Using a priors file has the advantage of allowing priors to be stored and edited seperately from the
												command lines.
											</p>
										
								</div>
								<div class="task">
										<div class="task_name">
												Getting nicer output
										</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -table-name Meta -cohort-names set1 set2
						-analysis-name "My first meta-analysis" 
					</div>
										<div class="task_notes">
												<b>Note</b>: Bingwa includes a number of options that help in giving more interpretable names
						to the output.	In the above command, we set the name of the table in which the primary output is produced
						(see the <a href="#file_formats">file formats page for details</a>), a name for the analysis overall,
						and names for each of the cohorts.
										</div>
								</div>
								<div class="task">
										<div class="task_name">
												Reweighting models
										</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -prior fixed:sd=0.2,0.2/cor=1,1,1 independent:sd=0.2,0.2/cor=1,0,1 -prior-weights fixed=1 independent=0.5
					</div>
										<div class="task_notes">
											<p>
												<b>Note</b>: By default all models are equally weighted in the model-average bayes factor.	Use the <code>-prior-weights</code>
						option to reweight them.	Weights are renormalised to sum to one in the mean bayes factor computation - i.e.
						the mean Bayes factor is computed as
						$$\text{BF}_{\text{avg}} = \frac{\sum_i \text{weight}_i \cdot \text{BF}_i}{\sum_i \text{weight}_i}$$
						where $\text{weight}_i$ and $\text{BF}_i$ are the specified weight and Bayes factor for model $i$.
						If <code>-prior-weights</code> is specified, any model not specified in the argument is given weight zero in the mean Bayes factor.
					</p>
					<p>
						The argument to <code>-prior-weights</code> can also be the name of a file, in which case bingwa will open the file
						and read prior weights from it (in the same format, i.e. <code>&lt;model name&gt;=&lt;weight&gt;</code> ).
					</p>
										</div>
								</div>
								<div class="task">
										<div class="task_name">
												Getting extra columns in the output
										</div>
										<div class="task_command_line">
												$ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -priors priors.txt -extra-columns controls_AA controls_AB controls_BB cases_AA cases_AB cases_BB frequentist_add_null_ll frequentist_add_alternative_ll
					</div>
										<div class="task_notes">
												<b>Note</b>: It's often useful to record extra information from the input files in the output.
						The <code>-extra-columns</code> option tells bingwa to read data from the specified columns in the input file
						and include it in the output.
										</div>
								</div>
				<div class="tutorial_separator"></div>
				<h3>
					Understanding bingwa sqlite output files
				</h3>
				<div>
					<p>
						Below we've given some examples of using the shell or various programming languages to extract data from a bingwa
						<a href="http://www.sqlite.org">sqlite</a> file.
						We assume the file is named <code>bingwa.sqlite</code> and the meta-analysis results were stored
						using the default prefix ('Bingwa') for table names.
					</p>
					<p>
											<b>Note</b>: The <code>sqlite3</code> program is installed by default on most UNIX-like systems.
						It's quite flexible, and can be used in interactive mode, or to run queries directly from the command-line.
						We'll use both methods below.
						See the <a href="https://www.sqlite.org/cli.html">sqlite3 command shell documentation</a> for a full list of options.
					</p>
					<h4>Inspecting the contents of the results file</h4>
					<p>
						sqlite files can be inspected using the sqlite3 command shell.	Open the file by typing:
					</p>
											<code>
							$ sqlite3 -column -header bingwa.sqlite <br/>
						</code>
					<p>
						You should see the sqlite prompt (<code>sqlite3&gt;</code>).	Let's see what tables are stored in this file:
					</p>
					<pre>
sqlite3&gt; .tables
Analysis							BingwaCounts					BingwaMetaView			
AnalysisProperty			BingwaCountsView			Variant							
AnalysisPropertyView	BingwaDetail					VariantIdentifier		
AnalysisStatus				BingwaDetailView			VariantView					
AnalysisStatusView		BingwaMeta					
					</pre>
										<p>
												Here, the main meta-analysis results are stored in the <code>BingwaMeta</code> table,
						and bingwa has also created the <code>BingwaMetaView</code> view which provides a nicer view of the results.
						The <code>BingwaCounts</code> and <code>BingwaDetail</code> tables contain, respectivelty,
						genotype counts taken from the input files, and the raw summary statistics (beta and standard error)
						that bingwa operated on.
					</p>
						<p>
							The <code>Analysis*</code> tables are used to store metadata about the analyses you have run.
						Let's see what's stored in this file:
					</p>
										<pre>
sqlite> SELECT * FROM Analysis ;
id					name						 chunk		 
----------	---------------	 ----------
1						bingwa analysis	 NA				 

sqlite> SELECT COUNT(*) AS count FROM BingwaMetaView
count	 
-------
200
					</pre>
					<p>
						So this file stores meta-analysis results for 200 variants from a single analysis called "bingwa analysis".
					</p>
					<p>
						In a large, real analysis, it's often useful to use the <code>-analysis-name</code> and <code>-analysis-chunk</code> options to give the analysis a 
						more descriptive name and explain what part of the data it corresponds to.	It's possible to store several analyses in the same sqlite file
						(or even in the same table, provided analysis columns are compatible), so this information becomes important to understand the results.
					</p>
					<p>
						Let's have a look at what the results actually look like.	 Because there are lots of columns we'll switch sqlite3 to line-based output:
					</p>
					<pre>
sqlite3&gt; .mode line
sqlite3&gt; SELECT * FROM BingwaMetaView LIMIT 1 ;
                                    rsid = RSID_2
                                  alleleA = A
                                  alleleB = -
                                 analysis = bingwa analysis
                              analysis_id = 1
                               variant_id = 2
                               chromosome = NA
                                 position = 2
                              cohort 1:bf = 0.685505148530258
                              cohort 2:bf = 0.754326805029774
   FixedEffectMetaAnalysis:included_betas = 11
                FixedEffectMetaAnalysis:N = 998.000188827515
FixedEffectMetaAnalysis:beta_1:add/bin2=1 = -0.048782749136173
             FixedEffectMetaAnalysis:se_1 = 0.130323119437487
    FixedEffectMetaAnalysis:wald_pvalue_1 = 0.708165117723267
           FixedEffectMetaAnalysis:pvalue = 0.708165117723267
                                 fixed:bf = 0.573458567964815
                           independent:bf = 0.517094908522291
                                  mean_bf = 0.545276738243553
                             max_bf_model = fixed
                                   max_bf = 0.573458567964815
                     best_posterior_model = fixed
                           best_posterior = 0.525841767807702
                 2nd_best_posterior_model = independent
                       2nd_best_posterior = 0.474158232192298
							</pre>
						<p>
						This analysis contains the results of both a fixed-effect frequentist and a bayesian analysis of two cohorts,
						by default called "cohort 1" and "cohort 2".	(The <code>-cohort-names</code> option could have been used to change these names).
					</p>
					<p>
						The first eight columns reflect metadata, that tells us the variant being tested and a link to the Analysis table (which useful if several analyses are stored in the same table).
						The chromosome information here is unspecified (the <code>-assume-chromosome</code> option could have been used
						to fill in chromosome information during the analysis).
					</p>
					<p>
						The next two columns contain Bayes factors (BFs) computed based on the data in each cohort independently.
						These are not strictly meta-analysis results, but are stored here for comparison with the meta-analysis Bayes factors
						below.  (This might change in a future version of the software.)
					</p>
					<p>
						The columns labelled 'FixedEffectMetaAnalysis' reflect the frequentist analysis.	They include 'N'
						a variable ('included_betas') which here indicates that data from both cohorts informed the analysis, and the
						combined effect size estimate and standard error from meta-analysis.	Here,	 because there is only one
						effect size being estimed, the two p-value columns (<code>pvalue</code> and <code>wald_pvalue_1</code>) are
						the same.
					</p>
					<p>
						The following columns reflect the bayesian meta-analysis.	 In order, they represent the BF for each model, 
						the model average BF, the maximum BF and the model with the max BF, details of the model with the highest posterior mass
						(which might be different from the maximum BF if the weighting is not uniform).	 Similarly, the last two columns reflect the model
						with the 2nd highest posterior mass.
					</p>
					<p>
						These particular results don't look very interesting (i.e. a p-value of <em>P</em>=0.7, or a Bayes factor less than one,
						suggest very little evidence that the indel is associated).  Nevertheless, if we
						were interested we could take a look at the data underlying this meta-analysis:
					</p>
					<pre>
sqlite3&gt; SELECT * FROM BingwaDetail WHERE analysis_id == 1  AND variant_id == 2;
               analysis_id = 1
                variant_id = 2
                chromosome = NA
                  position = 2
cohort 1:beta_1:add/bin2=1 = 0.0237287990748882
             cohort 1:se_1 = 0.186789005994797
           cohort 1:pvalue = 0.898887991905212
cohort 2:beta_1:add/bin2=1 = -0.117560997605324
             cohort 2:se_1 = 0.181916996836662
           cohort 2:pvalue = 0.517305016517639
</pre>
				 <p>
					 Estimated effects have opposite directions in the two cohorts.
					 Or we could look at the genotype counts that SNPTEST reported for this test:
				 </p>
				<pre>
sqlite3&gt; SELECT * FROM BingwaCounts WHERE analysis_id == 1  AND variant_id == 2;
                analysis_id = 1
                 variant_id = 2
                 chromosome = NA
                   position = 2
                 cohort 1:A = 0.0
                 cohort 1:B = 0.0
                cohort 1:AA = 24.1889991760254
                cohort 1:AB = 149.192001342773
                cohort 1:BB = 325.618988037109
              cohort 1:NULL = 3.98792002025139e-13
                 cohort 1:N = 498.999988555908
cohort 1:B_allele_frequency = 0.802034063901899
           cohort 1:trusted = 1
                 cohort 2:A = 0.0
                 cohort 2:B = 0.0
                cohort 2:AA = 28.2371997833252
                cohort 2:AB = 142.957000732422
                cohort 2:BB = 327.805999755859
              cohort 2:NULL = 5.11590986587707e-13
                 cohort 2:N = 499.000200271606
cohort 2:B_allele_frequency = 0.800169017777426
           cohort 2:trusted = 1
</pre>
				</div>
				<div class="tutorial_separator"></div>
				<h2>
					Processing bingwa sqlite output files
				</h2>
				<p>
					Below are some example commands that process bingwa output in the command-line or in R.
					See above for a full description of bingwa's sqlite output.
				</p>
					<div class="task">
							<div class="task_name">
									Extract a csv file of all results.
							</div>
							<div class="task_command_line">
									$ sqlite3 -header -csv bingwa.sqlite "SELECT * FROM BingwaMetaView" > my_file.csv
							</div>
					</div>
					<div class="task">
							<div class="task_name">
									Extract results from a specific region.
							</div>
							<div class="task_command_line">
									$ sqlite3 -header -csv bingwa.sqlite "SELECT * FROM BingwaMetaView WHERE chromosome == '05' AND position BETWEEN 120000000 AND 121000000" > my_file.csv
							</div>
							<div class="task_notes">
									<b>Note</b>: bingwa always stores chromosome identifiers as strings.	Currently these will be two-character strings with single-digit
			chromosomes appearing as <code>01</code>, <code>02</code>, etc.
							</div>
					</div>
					<div class="task">
							<div class="task_name">
									Read bingwa results into R
							</div>
							<div class="task_command_line">
			# R code to load bingwa results <br/>
									> library( RSQLite )<br/>
									> db = dbConnect( dbDriver( "SQLite" ), "bingwa.sqlite" )<br/>
									> D = dbGetQuery( db, "SELECT * FROM BingwaMetaView" )
							</div>
							<div class="task_notes">
									<b>Note</b>: You need the RSQLite library installed to process sqlite files from R.
							</div>
					</div>
				</div>
			</div>
			<div id="file_formats" style="display:none">
				<h3>Input file formats</h3>
								<div>
										Bingwa supports reading SNPTEST output files.	 Files are assumed space-, comma- or tab-delimited
								according to the file extension (<code>.txt</code>, <code>.csv</code> or <code>.tsv</code>).
								</div>
				<h3>Output file formats</h3>
				<p>
					Bingwa can either output to a flat file or to a table in a sqlite database.	 Both have pros and cons.
				</p>
				<p>
				<b>Flat files</b> are simple and easy to use.	 However, they can become difficult to read from for very large analyses
				(such as genome-wide analyses of datasets imputed into recent reference panels) which may require writing
				special file-handling code.
				</p>
				<p>
					Flat files are written when the <code>-flat-file</code> option is specified.
				</p>
				<p>
					<b>Sqlite files</b> are easy to use via the <a href="https://www.sqlite.org/cli.html">sqlite3 command shell</a>,
					which is installed on most UNIX-like systems.	 They can also be read directly from programming languages,
					e.g. by using the <a href="https://cran.r-project.org/package=RSQLite">RSQLite package</a> in <a href="">R</a> or
					the <a href="https://docs.python.org/2/library/sqlite3.html">sqlite3</a> module in python.
					An attractive feature is that they are <em>indexed</em>, so that data for specific regions of the genome or specific variants can
					be easily found.	See the <a href='#tutorial'>examples page</a> for code snippets for processing bingwa-produced sqlite files.
				</p>
				<p>Sqlite is the default output format for bingwa.</p>
				<h3>Specifying priors</h3>
				<p>
					The file format used for the <code>-priors</code> option follows the following rules.
					<ul>
						<li>Blank lines and lines starting with # (comment lines) are ignored.</li>
						<li>Every other line is part of a prior specification.</li>
						<li>Each prior must start with a name specification of the form '<code>&lt;name&gt;:</code>'.</li>
						<li>Two forms of prior specification are permissible, <em>full</em> and <em>shorthand</em> prior specification.</li>
						<li>
							A <em>full</em> prior specification is of the form <code>sd=a,b,c,.../cor=1,x,y,...</code>.
								Here <code>a,b,c,...,x,y,...</code> are real numbers.	 The number of sds specified must equal the combined number
								of parameters in all cohorts put together, denoted <em>d</em>; the number of correlations
								specified must then be equal to <code>d&times;(d+1)/2</code>.
								Conceptually, suppose the values <code>a,b,c,...</code> specify the diagonal entries of a matrix
								denoted &sigma;, and suppose the values <code>1,x,y,...</code> specify the upper triangle of
								the correlation matrix &Rho; (which should have 1s on the diagonal as in the example).
								The prior matrix used for analysis is then &sigma;&times;&Rho;&times;&sigma;.
						</li>
						<li>
							A <em>shorthand</em> prior specification is of the form <code>tau=z/sd=a,b,c,d,.../cor=1,x,y,...</code>.
								In this form, the number of sds specified must equal the number
								of parameters in each cohort (e.g. 1 for univariate analysis), denoted <em>d</em>;
								the number of correlations specified must then be equal to <code>d&times;(d+1)/2</code>.
						</li>
					</ul>
						</div>
						<div id="download" style="display:none">
								<div>
										<p>
												Bingwa is available as source code.
										</p>
										<h3>
												Binaries
										</h3>
										<p>
												Pre-compiled binaries are available for the following platforms.
						(See <a href="http://www.well.ox.ac.uk/~gav/resources">here</a> for other builds.)
										</p>
										<table>
												<tr>
														<th>
																Version
														</th>
														<th>
																Platform
														</th>
														<th>
																File
														</th>
												</tr>
												<tr>
														<td>v0.9-dev<sup>&dagger;</sup></td>
														<td>
																Ubuntu 12.04 <small>x86-64</small>
														</td>
														<td>
																<a href="http://www.well.ox.ac.uk/~gav/resources/bingwa_v0.9-dev-linux-x86_64.tgz">bingwa_v0.9-dev-linux-x86_64.tgz</a>
														</td>
												</tr>
												<tr>
														<td>v0.9-dev<sup>&dagger;</sup></td>
														<td>
																CentOS6.5 <small>x86-64</small>
														</td>
														<td>
																<a href="http://www.well.ox.ac.uk/~gav/resources/bingwa_v0.9-dev-linux-x86_64.tgz">bingwa_v0.9-dev-CentOS6.5-x86_64.tgz</a>
														</td>
												</tr>
												<tr>
														<td>v0.9-dev<sup>&dagger;</sup></td>
														<td>
																Mac OS X
														</td>
														<td>
																<a href="http://www.well.ox.ac.uk/~gav/resources/bingwa_v0.9-dev-osx.tgz">bingwa_v0.9-dev-osx.tgz</a>
														</td>
												</tr>
										</table>
										<p>
					<sup>&dagger;</sup><span class="smalltext">This version of bingwa is considered experimental.
					</p>
										<p>
												To run bingwa, download the relevant file and extract it as follows.
										</p>
										<pre>
$ tar -xzf bingwa_v0.9-dev-[machine].tgz
$ cd bingwa_v0.9-dev-[machine]
$ ./bingwa_v0.9-dev -help
</pre>
										<h3>
												Source
										</h3>
										<p>
												The source code to bingwa can be found on the <a href="http://bitbucket.org/gavinband/bingwa">bingwa page on bitbucket</a>.
										</p>
								</div>
						</div>
				</div>
		</body>
</html>
