<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
    <head>
        <link href="style/css/default.css" rel="stylesheet" type="text/css">

		<!-- Required scripts -->
        <script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
        <script type="text/javascript" src="js/jquery.ba-hashchange.min.js"></script>
        <script type="text/javascript" src="js/local.js"></script>

		<!-- MathJax configuration-->
		<script type="text/x-mathjax-config">
		  MathJax.Hub.Config({tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}});
		</script>        <!-- Google Analytics tracking code-->
		<script type="text/javascript"
		  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
		</script>

		<!-- Google analytics tracking code-->
        <script type="text/javascript">
            var _gaq = _gaq || [];
            _gaq.push(['_setAccount', 'UA-16521993-1']);
            _gaq.push(['_trackPageview']);

            (function() {
                var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
                ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
                var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
            })();
        </script>

		<!-- Go to first page if we get a vanilla URL-->
        <script type="text/javascript">
            $(document).ready(
                function() {
                    setFirstPage( "overview" ) ;
                }
            ) ;
        </script>
        <title></title>
    </head>
    <body>
        <div class="header" id="header1">
            <div class="header_text">
				bingwa
			</div>
            <div class="header_subtext">
				model-based meta-analysis for genome-wide association studies
            </div>
        </div>
        <div class="button_bar">
            <div class="nav_button" name="overview">
                <span class = "nav_button_content" >
                    overview
                </span>
            </div>
            <div class="nav_button" name="documentation">
                <span class = "nav_button_content" >
                    documentation
                </span>
            </div>
            <div class="nav_button" name="tutorial">
                <span class = "nav_button_content" >
                    examples
                </span>
            </div>
            <div class="nav_button" name="file_formats">
                <span class = "nav_button_content" >
                    file formats
                </span>
            </div>
            <div class="nav_button" name="download">
                <span class = "nav_button_content" >
                    download
                </span>
            </div>
        </div>
        <div class="main_text">
            <div id="overview" style="display:none">
                <p>
					<em>This page documents version 2.0-dev of bingwa, which is currently considered experimental</em>.
				</p>
                <p>
					<strong>Introduction.</strong>
					<span class="program_name">bingwa</span> is a command-line tool used to carry out meta-analysis of genome-wide association studies.
					It is particularly suited for use with the program <a href="https://mathgen.stats.ox.ac.uk/genetics_software/snptest/snptest.html">SNPTEST</a>,
					but can also be used with other association testing programs such as <a href="http://www.helsinki.fi/~mjxpirin/download.html">MMM</a>.
					Features of bingwa include:
				</p>
				<ul>
					<li>
						Direct support for SNPTEST output files.
					</li>
					<li>
						Support for fixed-effect frequentist and a flexible bayesian meta-analysis method.
					</li>
					<li>
						Support for both univariate and multivariate meta-analysis (including support
						for meta-analysing tests performed using  <a href="https://mathgen.stats.ox.ac.uk/genetics_software/snptest/snptest.html#multinomial_tests">SNPTEST's multinomial test</a>).
					</li>
				</ul>
                <p>
                    <strong>Disclaimer.</strong>
					This page documents version 2.0-dev of bingwa which is currently considered <b>experimental</b>.
					This means we expect some features not to work, or
					not to work well, or to work wrongly, or to destroy your computer or sanity.
                </p>
                <p>
					If you make use of bingwa, please cite the following paper, which describes the key methodology underlying the program:
					<ul>
					<li>
						"<em>Imputation-Based Meta-Analysis of Severe Malaria in Three African Populations</em>",
						Band et al, PLOS Genetics (2013)
					</li>
				</ul>
				</p>
				<p>
					Similar methodology was employed in these papers:
				<ul>
					<li>
						"<em>Genome-wide association study identifies a variant in HDAC9 associated with large vessel ischemic stroke</em>",
						International Stroke Genetics Consortium (ISGC) and Wellcome Trust Case Control Consortium 2 (WTCCC2), Nat. Genetics (2012)
					</li>
					<li>"<em>Reappraisal of known malaria resistance loci in a large multi-centre study</em>", Rockett et al, Nat. Genetics (2014).</li>
				</ul>
                <p>
                    <strong>Acknowledgements.</strong>  Methodological work underpinning the use of bingwa was
					carried out by Matti Pirinen, Chris Spencer, Gavin Band and others at
					the <a href="http://www.well.ox.ac.uk/home">Wellcome Trust Centre for Human Genetics</a>, Oxford, UK.
					This software was implemented by Gavin Band.
                </p>
            </div>
            <div id="documentation" style="display:none">
                <div class="display_diagram_right">
                    <!--<object data="style/images/inthinnerator-flow.svg"  type="image/svg+xml" align="center">
					</object>-->
                </div>
                <p>
                    The general process followed by bingwa is shown on the right.  In brief, bingwa proceeds by
					<ol>
						<li>Loading association test data from a set of cohorts (e.g. SNPTEST files).</li>
						<li>Matching variants between cohorts.</li>
						<li>Computing frequentist meta-analysis and one or more Bayes factors.</li>
						<li>Writing results.</li>
					</ol>
                </p>
				<p>
					We'll update this page with more details on the method soon.  In the meantime the <a href="#tutorial">examples page</a>
					has usage examples.
                </p>
			</div>
            <div id="tutorial" style="display:none">
				<h2>
					Running bingwa
				</h2>
                <div class="tutorial_header">
                    This page shows common command lines to use bingwa.  These examples assume the program is being run from a directory
					containing two SNPTEST files, <span class="monospace">cohort1.snptest</span> and <span class="monospace">cohort2.snptest</span>.
                </div>
                <div class="tutorial_separator">
                </div>
                <div class="task">
                    <div class="task_name">
                        View the program usage page
                    </div>
                    <div class="task_command_line">
                        $ bingwa -help | less
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Compute fixed-effect meta-analysis of the two files
                    </div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite
					</div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -flat-file -o output.csv
                    </div>
                    <div class="task_notes">
                        <b>Note</b>: By default bingwa writes output to a sqlite file (more details below).  Use <code>-flat-file</code> to write a flat file instead.
							Flat files are space, tab or comma-delimited according to the filename extension.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Compute a fixed-effect bayesian meta-analysis
                    </div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -prior sd=0.2,0.2/cor=1,1,1 
					</div>
                    <div class="task_notes">
                        <b>Note</b>: The above specifies a multivariate normal (MVN) prior on the 'true' effects
						that is centred at zero and has variance-covariance matrix with every entry equal 0.2<sup>2</sup>.
						(Bingwa will print this model to the screen output so you can double-check this).
						Under this model the true effects in the two populations are assumed perfectly correlated; the only
						variation comes from estimation effects represented by the within-study standard errors.
						The bayes factor for this model will be stored in a column named <code>sd=0.2,0.2/cor=1,1,1:bf</code>.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Compute both a fixed and independent-effect bayesian meta-analysis
                    </div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -prior fixed:sd=0.2,0.2/cor=1,1,1 -prior independent:sd=0.2,0.2/cor=1,0,1 
					</div>
                    <div class="task_notes">
                        <b>Note</b>: In the above command we've named the models to avoid confusion - now the
						output will contain <code>fixed:bf</code> and <code>independent:bf</code> columns.
						Here, the independent model assumes that true effects are uncorrelated between the cohorts - i.e.
						they have diagonal variance-covariance matrix.
                        In addition to the two model-specific bayes factors, bingwa will compute a model-averaged bayes factor
						over all the models specified, named <code>mean_bf</code>.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Avoiding typing all those sds and cors.
                    </div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest ... cohort&lt;n&gt;.snptest -o output.sqlite -prior fixed:tau=1/sd=0.2/cor=1 -prior independent:tau=0/sd=0.2/cor=1
					</div>
                    <div class="task_notes">
                        <b>Note</b>: In the above command we use a short-hand version of the prior specification.
						The first part (tau) denotes the between-cohort correlation.  The rest now denotes the prior
						covariance matrix within each cohort.  Thus as before the fixed-effect model reflects a prior
						covariance with all entries equal to 0.2<sup>2</sup>, while the independent-effect model
						reflects a diagonal covariance matrix.
                    </div>
					<div class="task_notes">
						We advise always checking the screen output to debug model definitions.
					</div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Getting nicer output
                    </div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -table-name Meta -cohort-names set1 set2
						-analysis-name "My first meta-analysis" 
					</div>
                    <div class="task_notes">
                        <b>Note</b>: Bingwa includes a number of options that help in giving more interpretable names
						to the output.  In the above command, we set the name of the table in which the primary output is produced
						(see the <a href="#file_formats">file formats page for details</a>), a name for the analysis overall,
						and names for each of the cohorts.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Reweighting models
                    </div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite -prior fixed:sd=0.2,0.2/cor=1,1,1 independent:sd=0.2,0.2/cor=1,0,1 -prior-weights fixed=1 independent=0.5
					</div>
                    <div class="task_notes">
                        <b>Note</b>: By default all models are equally weighted in the model-average bayes factor.  Use the <code>-prior-weights</code>
						option to reweight them.  Weights are renormalised to sum to one in the mean bayes factor computation - i.e. the mean Bayes factor is computed as
						$$\text{BF}_{\text{avg}} = \frac{\sum_i \text{weight}_i \cdot \text{BF}_i}{\sum_i \text{weight}_i}$$
						where $\text{weight}_i$ and $\text{BF}_i$ are the specified weight and Bayes factor for model $i$.
                    </div>
                </div>
				<div class="tutorial_separator"></div>
				<h3>
					Understanding bingwa sqlite output files
				</h3>
				<div>
					<p>
						Below we've given some examples of using the shell or various programming languages to extract data from a bingwa
						<a href="http://www.sqlite.org">sqlite</a> file.
						We assume the file is named <code>bingwa.sqlite</code> and the meta-analysis results were stored
						in the <code>MetaAnalysis</code> table.
					</p>
					<p>
	                    <b>Note</b>: The <code>sqlite3</code> program is installed by default on most UNIX-like systems.
						It's quite flexible, and can be used in interactive mode, or to run queries directly from the command-line.
						We'll use both methods below.
						See the <a href="https://www.sqlite.org/cli.html">sqlite3 command shell documentation</a> for a full list of options.
					</p>
					<h4>Inspecting the contents of the results file</h4>
					<p>
						sqlite files can be inspected using the sqlite3 command shell.  Open the file by typing:
					</p>
	                    <code>
							$ sqlite3 -column -header bingwa.sqlite <br/>
						</code>
					<p>
						You should see the sqlite prompt (<code>sqlite3&gt;</code>).  Let's see what tables are stored in this file:
					</p>
					<pre>
sqlite3&gt; .tables
Analysis                Entity                  MetaAnalysis          
AnalysisProperty        EntityData              MetaAnalysisView      
AnalysisPropertyView    EntityDataView          Variant               
AnalysisStatus          EntityRelationship      VariantIdentifier     
AnalysisStatusView      EntityRelationshipView  VariantView
					</pre>
                    <p>
                        Here, the main results are stored in the <code>MetaAnalysis</code> table,
						and bingwa has also created the <code>MetaAnalysisView</code> view which provides a nicer view of the results.
						(The <code>-table-name</code> option can be used to change the name of the main results table.
						By default it is a name of the form <code>Analysis&lt;n&gt;</code>, where n is an integer.)
						The <code>Analysis*</code> and <code>Entity*</code> tables are used to store
						metadata about the analyses you have run.  Let's see what's stored in this file:
					</p>
                    <pre>
sqlite> SELECT * FROM Analysis ;
id          name             chunk     
----------  ---------------  ----------
1           bingwa analysis  NA        

sqlite> SELECT COUNT(*) AS count FROM MetaAnalysisView
count  
-------
200
					</pre>
					<p>
						So this file stores meta-analysis results for 200 variants from a single analysis called "bingwa analysis".
					</p>
					<p>
						In a large, real analysis, it's often useful to use the <code>-analysis-name</code> and <code>-analysis-chunk</code> options to give the analysis a 
						more descriptive name and explain what part of the data it corresponds to.  It's possible to store several analyses in the same sqlite file
						(or even in the same table, provided analysis columns are compatible), so this information becomes important to understand the results.
					</p>
					<p>
						Let's have a look at what the results actually look like.  Because there are lots of columns we'll switch sqlite3 to line-based output:
					</p>
					<pre>
sqlite3&gt; .mode line
sqlite3&gt; SELECT * FROM MetaAnalysisView LIMIT 1 ;
                                             rsid = KGP1
                                          alleleA = A
                                          alleleB = G
                                         analysis = bingwa analysis
                                      analysis_id = 1
                                       variant_id = 1
                                       chromosome = NA
                                         position = 1000
                                       cohort 1:A = 0.0
                                       cohort 1:B = 0.0
                                      cohort 1:AA = 168.889999389648
                                      cohort 1:AB = 240.240005493164
                                      cohort 1:BB = 87.870002746582
                                    cohort 1:NULL = 0.0
                      cohort 1:B_allele_frequency = 0.418490950302478
                                     cohort 1:maf = 0.418489992618561
                               cohort 1:ml_beta_1 = 0.100000001490116
                                 cohort 1:ml_se_1 = 0.0199999995529652
                                  cohort 1:pvalue = 0.100000001490116
                                    cohort 1:info = 0.374507993459702
                                       cohort 2:A = 0.0
                                       cohort 2:B = 0.0
                                      cohort 2:AA = 176.800003051758
                                      cohort 2:AB = 250.979995727539
                                      cohort 2:BB = 71.2190017700195
                                    cohort 2:NULL = 0.0
                      cohort 2:B_allele_frequency = 0.394207201652197
                                     cohort 2:maf = 0.394199997186661
                               cohort 2:ml_beta_1 = 0.0799999982118607
                                 cohort 2:ml_se_1 = 0.0199999995529652
                                  cohort 2:pvalue = 0.94200998544693
                                    cohort 2:info = 0.97086101770401
                   FixedEffectMetaAnalysis:beta_1 = 0.0899999998509884
                     FixedEffectMetaAnalysis:se_1 = 0.0141421353076296
            FixedEffectMetaAnalysis:wald_pvalue_1 = 1.96615875463349e-10
                   FixedEffectMetaAnalysis:pvalue = 1.96615875463348e-10
			        </pre>
			   	 	<p>
						This analysis contains the results of a fixed-effect frequentist analysis of two cohorts, by default called "cohort 1" and "cohort 2".
						The results file contains columns reflecting genotype counts and model fits in each cohort, and columns containing meta-analysed
						effect size estimate, standard error, and a two-tailed fixed-effect meta-analysis P-value.
					</p>
					<p>
						<b>Note</b>: the <code>-cohort-names</code> option can be used to name the cohorts in a more useful way.
					</p>
				</div>
                <div class="tutorial_separator"></div>
				<h2>
					Processing bingwa sqlite output files
				</h2>
				<p>
					Below are some example commands that process bingwa output in the command-line or in R.
					See above for a full description of bingwa's sqlite output.
				</p>
                <div class="task">
                    <div class="task_name">
                        Extract a csv file of all results.
                    </div>
                    <div class="task_command_line">
                        $ sqlite3 -header -csv bingwa.sqlite "SELECT * FROM MetaAnalysisView" > my_file.csv
					</div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Extract results from a specific region.
                    </div>
                    <div class="task_command_line">
                        $ sqlite3 -header -csv bingwa.sqlite "SELECT * FROM MetaAnalysisView WHERE chromosome == '05' AND position BETWEEN 120000000 AND 121000000" > my_file.csv
					</div>
                    <div class="task_notes">
                        <b>Note</b>: bingwa always stores chromosome identifiers as strings.  Currently these will be two-character strings with single-digit
						chromosomes appearing as <code>01</code>, <code>02</code>, etc.
                    </div>
					
                </div>
                <div class="task">
                    <div class="task_name">
                        Read bingwa results into R
                    </div>
                    <div class="task_command_line">
						# R code to load bingwa results <br/>
                        > library( RSQLite )<br/>
                        > db = dbConnect( dbDriver( "SQLite" ), "bingwa.sqlite" )<br/>
                        > D = dbGetQuery( db, "SELECT * FROM MetaAnalysisView" )
					</div>
                    <div class="task_notes">
                        <b>Note</b>: You need the RSQLite library installed to process sqlite files from R.
                    </div>
					
                </div>
            </div>
            <div id="file_formats" style="display:none">
				<h3>Input file formats</h3>
                <div>
                    Bingwa supports reading SNPTEST output files.  Files are assumed space-, comma- or tab-delimited
								according to the file extension (<code>.txt</code>, <code>.csv</code> or <code>.tsv</code>).
                </div>
				<h3>Output file formats</h3>
				<p>
					Bingwa can either output to a flat file or to a table in a sqlite database.  Both have pros and cons.
				</p>
				<p>
				<b>Flat files</b> are simple and easy to use.  However, they can become difficult to read from for very large analyses
				(such as genome-wide analyses of datasets imputed into recent reference panels) which may require writing
				special file-handling code.
				</p>
				<p>
					Flat files are written when the <code>-flat-file</code> option is specified.
				</p>
				<p>
					<b><a href="http://www.sqlite.org">Sqlite</a> files</b> are also easy to use via the <a href="https://www.sqlite.org/cli.html">sqlite3 command shell</a>,
					which is installed on most UNIX-like systems.  They can also be read directly from programming languages such as <a href="">R</a>.
					An attractive feature is that they are <em>indexed</em>, so that data for specific regions of the genome or specific variants can
					be easily found.  See the <a href='#tutorial'>examples page</a> for code snippets for processing bingwa-produced sqlite files.
				</p>
				<p>Sqlite is the default output format for bingwa.</p>
            </div>
            <div id="download" style="display:none">
                <div>
                    <p>
                        Bingwa is available either as binaries or as source code.
                    </p>
                    <h3>
                        Binaries
                    </h3>
                    <p>
                        Pre-compiled binaries are available for the following platforms.
						(See <a href="http://www.well.ox.ac.uk/~gav/resources">here</a> for other builds.)
                    </p>
                    <table>
                        <tr>
                            <th>
                                Version
                            </th>
                            <th>
                                Platform
                            </th>
                            <th>
                                File
                            </th>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                Ubuntu 12.04 <small>x86-64</small>
                            </td>
                            <td>
                                <a href="http://www.well.ox.ac.uk/~gav/resources/bingwa_v2.0-dev-linux-x86_64.tgz">bingwa_v2.0-dev-linux-x86_64.tgz</a>
                            </td>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                CentOS6.5 <small>x86-64</small>
                            </td>
                            <td>
                                <a href="http://www.well.ox.ac.uk/~gav/resources/bingwa_v2.0-dev-linux-x86_64.tgz">bingwa_v2.0-dev-CentOS6.5-x86_64.tgz</a>
                            </td>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                Mac OS X
                            </td>
                            <td>
                                <a href="http://www.well.ox.ac.uk/~gav/resources/bingwa_v2.0-dev-osx.tgz">bingwa_v2.0-dev-osx.tgz</a>
                            </td>
                        </tr>
                    </table>
                    <p>
					<sup>&dagger;</sup><span class="smalltext">This version of bingwa is considered experimental.
					</p>
                    <p>
                        To run bingwa, download the relevant file and extract it as follows.
                    </p>
                    <pre>
$ tar -xzf bingwa_v2.0-dev-[machine].tgz
$ cd bingwa_v2.0-dev-[machine]
$ ./bingwa_v2.0-dev -help
</pre>
                    <h3>
                        Source
                    </h3>
                    <p>
                        The source code to bingwa is built as part of the <a href="http://bitbucket.org/gavinband/qctool">qctool package on bitbucket</a>.
						See <a href="https://bitbucket.org/gavinband/qctool/wiki/Home">the qctool wiki page</a> for more details.
					</p>
                </div>
            </div>
        </div>
    </body>
</html>
