<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
    <head>
        <script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
        <script type="text/javascript" src="js/jquery.ba-hashchange.min.js"></script>
        <script type="text/javascript" src="js/local.js"></script>

        <!-- Google Analytics tracking code-->
        <script type="text/javascript">
            var _gaq = _gaq || [];
            _gaq.push(['_setAccount', 'UA-16521993-1']);
            _gaq.push(['_trackPageview']);

            (function() {
                var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
                ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
                var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
            })();
        </script>

        <script type="text/javascript">
            $(document).ready(
                function() {
                    setFirstPage( "overview" ) ;
                }
            ) ;
        </script>
        <link href="style/css/default.css" rel="stylesheet" type="text/css">
        <title></title>
    </head>
    <body>
        <div class="header" id="header1">
            <div class="header_text">
				bingwa
			</div>
            <div class="header_subtext">
				<b>b</b>ayes<b>i</b>a<b>n</b> meta-analysis for <b>g</b>enome-<b>w</b>ide <b>a</b>ssociation studies
            </div>
        </div>
        <div class="button_bar">
            <div class="nav_button" name="overview">
                <span class = "nav_button_content" >
                    overview
                </span>
            </div>
            <div class="nav_button" name="documentation">
                <span class = "nav_button_content" >
                    documentation
                </span>
            </div>
            <div class="nav_button" name="tutorial">
                <span class = "nav_button_content" >
                    examples
                </span>
            </div>
            <div class="nav_button" name="file_formats">
                <span class = "nav_button_content" >
                    file formats
                </span>
            </div>
            <div class="nav_button" name="download">
                <span class = "nav_button_content" >
                    download
                </span>
            </div>
        </div>
        <div class="main_text">
            <div id="overview" style="display:none">
                <p>
					<strong>Introduction.</strong>
					<b>b</b>ingwa is a command-line tool used to carry out meta-analysis of genome-wide association studies.
					It is particularly suited for use with the program <a href="">SNPTEST</a>, but can also be used with other programs such as <a href="">MMM</a>.
					Key features of bingwa include:
				</p>
				<ul>
					<li>
						Computation of fixed-effect frequentist and flexible bayesian meta-analysis.
					</li>
					<li>
						Support 
					</li>
				</ul>
                <p>
                    <strong>Change history.</strong>
					<br />
					<em>Important!</em>
					This page documents version 2 of bingwa which is currently <b>experimental</b>.  This means we expect some features not to work, or
					not to work well, or to work wrongly, or to destroy your computer or sanity.
                </p>
                <p>
                    <strong>Acknowledgements.</strong> The following people contributed to the methods, design and implementation of this software:
                </p>
                <p>
                    <ul>
                        <li>
                            <a href="http://www.well.ox.ac.uk/~gav/">Gavin Band</a>
                        </li>
                        <li>
                            <a href="http://www.well.ox.ac.uk/chris-spencer">Chris Spencer</a>
                        </li>
                        <li>
                            <a href="http://www.well.ox.ac.uk/si-quang-le-2">Quang Si Le</a>
                        </li>
                        <li>
                            <a href="http://www.helsinki.fi/~mjxpirin/">Matti Pirinen</a>
                        </li>
                        <li>
                            <a href="http://www.well.ox.ac.uk/geraldine-clarke">Geraldine Clarke</a>
                        </li>
					</ul>
                </p>
                <p>
                    <strong>Contact.</strong> For more information or questions, please contact the oxstatgen mailing list at
                    <pre class="monospace">
                        oxstatgen (at) jiscmail.ac.uk
                    </pre>
                </p>
            </div>
            <div id="documentation" style="display:none">
                <div class="display_diagram_right">
                    <!--<object data="style/images/inthinnerator-flow.svg"  type="image/svg+xml" align="center">
					</object>-->
                </div>
                <p>
                    The general process followed by bingwa is shown on the right.  In brief, bingwa proceeds by
					<ol>
						<li>Loading association test data from a set of cohorts (e.g. SNPTEST files).</li>
						<li>Matching variants between cohorts.</li>
						<li>Computing frequentist meta-analysis and one or more Bayes factors.</li>
						<li>Writing results.</li>
					</ol>
                </p>
			</div>
            <div id="tutorial" style="display:none">
				<h2>
					Running bingwa
				</h2>
                <div class="tutorial_header">
                    This page shows common command lines to use bingwa.  These examples assume the program is being run from a directory
					containing two SNPTEST files, <span class="monospace">cohort1.snptest</span> and <span class="monospace">cohort2.snptest</span>.
                </div>
                <div class="tutorial_separator">
                </div>
                <div class="task">
                    <div class="task_name">
                        View the program usage page
                    </div>
                    <div class="task_command_line">
                        $ bingwa -help | less
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Compute fixed-effect meta-analysis of the two files
                    </div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -o output.sqlite
					</div>
                    <div class="task_command_line">
                        $ bingwa -data cohort1.snptest cohort2.snptest -flat-file -o output.csv
                    </div>
                    <div class="task_notes">
                        <b>Note</b>: By default bingwa writes output to a sqlite file.  Use <code>-flat-file</code> to write a flat file instead.
							Flat files are space, tab or comma-delimited according to the filename extension.
                    </div>
                </div>
                <div class="tutorial_separator">
                </div>
				<h2>
					Understanding bingwa sqlite output files
				</h2>
				<div>
					<p>
						Below we've given some examples of using the shell or various programming languages to extract data from a bingwa sqlite file.
						We assume the file is named <code>bingwa.sqlite</code> and the meta-analysis results were stored
						in the <code>MetaAnalysis</code> table.
					</p>
					<p>
	                    <b>Note</b>: The <code>sqlite3</code> program is installed by default on most UNIX-like systems.
						It's quite flexible.  It can be used in interactive mode, or to run queries directly from the command-line.
						See the <a href="https://www.sqlite.org/cli.html">sqlite3 command shell documentation</a> for a full list of options.
					</p>
					<h3>Inspecting the contents of the results file</h3>
					<p>
						sqlite files can be inspected using the sqlite3 command shell.  Open the file by typing:
					</p>
	                    <code>
							$ sqlite3 -column -header bingwa.sqlite <br/>
						</code>
					<p>
						You should see the sqlite prompt (<code>sqlite3&gt;</code>).  Let's see what tables are stored in this file:
					</p>
					<pre>
sqlite3&gt; .tables
Analysis                Entity                  MetaAnalysis          
AnalysisProperty        EntityData              MetaAnalysisView      
AnalysisPropertyView    EntityDataView          Variant               
AnalysisStatus          EntityRelationship      VariantIdentifier     
AnalysisStatusView      EntityRelationshipView  VariantView
					</pre>
                    <p>
                        Here, the main results are here stored in the <code>MetaAnalysis</code> table,
						and bingwa has also created the <code>MetaAnalysisView</code> view which provides a nicer view of the results.
						(The <code>-table-name</code> option can be used to change the name of the main results table.  By default it is a name of the form Analysis&lt;n&gt;, where n is an integer.)
						The <code>Analysis*</code> and <code>Entity*</code> tables are used to store
						metadata about the analyses you have run.  Let's see what's stored in this file:
					</p>
                    <pre>
sqlite> SELECT * FROM Analysis ;
id          name             chunk     
----------  ---------------  ----------
1           bingwa analysis  NA        

sqlite> SELECT COUNT(*) AS count FROM MetaAnalysisView
count  
-------
200
					</pre>
					<p>
						So this file stores meta-analysis results for 200 variants from a single analysis called "bingwa analysis".
					</p>
					<p>
						In a large, real analysis, it's often useful to use the <code>-analysis-name</code> and <code>-analysis-chunk</code> options to give the analysis a 
						more descriptive name and explain what part of the data it corresponds to.  It's possible to store several analyses in the same sqlite file
						(or even in the same table, provided analysis columns are compatible), so this information becomes important to understand the results.
					</p>
					<p>
						Let's have a look at what the results actually look like.  Because there are lots of columns we'll switch sqlite3 to line-based output:
					</p>
					<pre>
sqlite3&gt; .mode line
sqlite3&gt; SELECT * FROM MetaAnalysisView LIMIT 1 ;
                                             rsid = KGP1
                                          alleleA = A
                                          alleleB = G
                                         analysis = bingwa analysis
                                      analysis_id = 1
                                       variant_id = 1
                                       chromosome = NA
                                         position = 1000
                                       cohort 1:A = 0.0
                                       cohort 1:B = 0.0
                                      cohort 1:AA = 168.889999389648
                                      cohort 1:AB = 240.240005493164
                                      cohort 1:BB = 87.870002746582
                                    cohort 1:NULL = 0.0
                      cohort 1:B_allele_frequency = 0.418490950302478
                                     cohort 1:maf = 0.418489992618561
                               cohort 1:ml_beta_1 = 0.100000001490116
                                 cohort 1:ml_se_1 = 0.0199999995529652
                                  cohort 1:pvalue = 0.100000001490116
                                    cohort 1:info = 0.374507993459702
                                       cohort 2:A = 0.0
                                       cohort 2:B = 0.0
                                      cohort 2:AA = 176.800003051758
                                      cohort 2:AB = 250.979995727539
                                      cohort 2:BB = 71.2190017700195
                                    cohort 2:NULL = 0.0
                      cohort 2:B_allele_frequency = 0.394207201652197
                                     cohort 2:maf = 0.394199997186661
                               cohort 2:ml_beta_1 = 0.0799999982118607
                                 cohort 2:ml_se_1 = 0.0199999995529652
                                  cohort 2:pvalue = 0.94200998544693
                                    cohort 2:info = 0.97086101770401
                   FixedEffectMetaAnalysis:beta_1 = 0.0899999998509884
                     FixedEffectMetaAnalysis:se_1 = 0.0141421353076296
            FixedEffectMetaAnalysis:wald_pvalue_1 = 1.96615875463349e-10
                   FixedEffectMetaAnalysis:pvalue = 1.96615875463348e-10
			        </pre>
			   	 	<p>
						This analysis contains the results of a fixed-effect frequentist analysis of two cohorts, by default called "cohort 1" and "cohort 2".
						The results file contains columns reflecting genotype counts and model fits in each cohort, and columns containing meta-analysed
						effect size estimate, standard error, and a two-tailed fixed-effect meta-analysis P-value.
					</p>
					<p>
						<b>Note</b>: the <code>-cohort-names</code> option can be used to name the cohorts in a more useful way.
					</p>
				</div>
			<div>
				<h2>
					Processing bingwa sqlite output files
				</h2>
				<p>
					Below are some example commands that process bingwa output in the command-line or in R.
				</p>
                <div class="task">
                    <div class="task_name">
                        Extract a csv file of all results.
                    </div>
                    <div class="task_command_line">
                        $ sqlite3 -header -csv bingwa.sqlite "SELECT * FROM MetaAnalysisView" > my_file.csv
					</div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Extract results from a specific region.
                    </div>
                    <div class="task_command_line">
                        $ sqlite3 -header -csv bingwa.sqlite "SELECT * FROM MetaAnalysisView WHERE chromosome == '05' AND position BETWEEN 120000000 AND 121000000" > my_file.csv
					</div>
                    <div class="task_notes">
                        <b>Note</b>: bingwa always stores chromosome identifiers as strings.  Currently these will be two-character strings with single-digit
						chromosomes appearing as <code>01</code>, <code>02</code>, etc.
                    </div>
					
                </div>
                <div class="task">
                    <div class="task_name">
                        Read bingwa results into R
                    </div>
                    <div class="task_command_line">
						# R code to load bingwa results <br/>
                        > library( RSQLite )<br/>
                        > db = dbConnect( dbDriver( "SQLite" ), "bingwa.sqlite" )<br/>
                        > D = dbGetQuery( db, "SELECT * FROM MetaAnalysisView" )
					</div>
                    <div class="task_notes">
                        <b>Note</b>: You need the RSQLite library installed to process sqlite files from R.
                    </div>
					
                </div>
            </div>
		</div>
            <div id="file_formats" style="display:none">
				<h3>Input file formats</h3>
                <div>
                    Bingwa supports the following input file formats:
                    <table class="file_formats">
                        <tr>
                            <th>
                                format
                            </th>
                            <th>
                                recognised extension(s)
                            </th>
                            <th>
                                notes
                            </th>
                        </tr>
                        <tr>
                            <td>
                                <a href="https://mathgen.stats.ox.ac.uk/genetics_software/snptest/snptest.html">SNPTEST</a> output file
                            </td>
                            <td>
								(none)
                            </td>
                            <td>
								These can be supplied to the <span class="monospace">-data</span> option.  Files are space-, comma- or tab-delimited
								according to the file extension (<code>.txt</code>, <code>.csv</code> or <code>.tsv</code>)
                            </td>
                        </tr>
                    </table>
                </div>
				<h3>Output file formats</h3>
				<div>
				<p>Bingwa can either output to a flat file or to a table in a sqlite database.  Both have pros and cons.</p>
				<p>
				<b>Flat files</b> are simple and easy to use.  However, they can become difficult to read from for very large analyses
				(such as genome-wide analyses of datasets imputed into recent reference panels) which may require writing
				special file-handling code.
				</p>
				<p>
					<b><a href="http://www.sqlite.org">Sqlite</a> files</b> are also easy to use via the <a href="">sqlite3</a> command-line client,which 
					is installed on most UNIX-like systems.  They can also be read directly from programming languages such as <a href="">R</a>.
					An attractive feature is that they are <em>indexed</em>, so that data for specific regions of the genome or specific variants can
					be easily found.  See the <a href='#tutorial'>examples page</a> for code snippets for processing bingwa-produced sqlite files.
				</p>
			</div>
            </div>
            <div id="download" style="display:none">
                <div>
                    <p>
                        Bingwa is available either as binaries or as source code.
                    </p>
                    <h3>
                        Binaries
                    </h3>
                    <p>
                        Pre-compiled binaries are available for the following platforms.
                    </p>
                    <table>
                        <tr>
                            <th>
                                Version
                            </th>
                            <th>
                                Platform
                            </th>
                            <th>
                                File
                            </th>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                Linux <small>x86-64 static build</small>
                            </td>
                            <td>
                                <a href="resources/bingwa_v2.0-dev-linux-x86_64.tgz">inthinnerator_v2.0-dev-linux-x86_64.tgz</a>
                            </td>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                Mac OS X
                            </td>
                            <td>
                                <a href="resources/bingwa_v2.0-dev-osx.tgz">inthinnerator_v2.0-dev-osx.tgz</a>
                            </td>
                        </tr>
                    </table>
                    <p>
					<sup>&dagger;</sup><span class="smalltext">This version of bingwa is considered experimental.
					</p>
                    <p>
                        To run bingwa, download the relevant file and extract it as follows.
                    </p>
                    <pre>
$ tar -xzf bingwa_v2.0-dev-[machine].tgz
$ cd bingwa_v2.0-dev-[machine]
$ ./bingwa_v2.0-dev -help
</pre>
                    <h3>
                        Source
                    </h3>
                    <p>
                        The source code to bingwa is available as part of the <a href="http://bitbucket.org/gavinband/qctool">qctool package on bitbucket</a>.
						See <a href="http://www.well.ox.ac.uk/~gav/qctool#download">this page</a> for details.
					</p>
                </div>
            </div>
        </div>
    </body>
</html>
