<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
    <head>
        <script type="text/javascript" src="js/jquery-1.4.2.min.js"></script>
        <script type="text/javascript" src="js/jquery.ba-hashchange.min.js"></script>
        <script type="text/javascript" src="js/local.js"></script>

        <!-- Google Analytics tracking code-->
        <script type="text/javascript">
            var _gaq = _gaq || [];
            _gaq.push(['_setAccount', 'UA-16521993-1']);
            _gaq.push(['_trackPageview']);

            (function() {
                var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
                ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
                var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
            })();
        </script>

        <script type="text/javascript">
            $(document).ready(
                function() {
                    setFirstPage( "overview" ) ;
                }
            ) ;
        </script>
        <link href="style/css/default.css" rel="stylesheet" type="text/css">
        <title></title>
    </head>
    <body>
        <div class="header" id="header1">
            <div class="header_text">
				bingwa
			</div>
            <div class="header_subtext">
				<b>b</b>ayes<b>i</b>an meta-a<b>n</b>alysis for <b>g</b>enome-<b>w</b>ide <b>a</b>ssociation studies
            </div>
        </div>
        <div class="button_bar">
            <div class="nav_button" name="overview">
                <span class = "nav_button_content" >
                    overview
                </span>
            </div>
            <div class="nav_button" name="documentation">
                <span class = "nav_button_content" >
                    documentation
                </span>
            </div>
            <div class="nav_button" name="tutorial">
                <span class = "nav_button_content" >
                    examples
                </span>
            </div>
            <div class="nav_button" name="file_formats">
                <span class = "nav_button_content" >
                    file formats
                </span>
            </div>
            <div class="nav_button" name="download">
                <span class = "nav_button_content" >
                    download
                </span>
            </div>
        </div>
        <div class="main_text">
            <div id="overview" style="display:none">
                <p>
					<strong>Introduction.</strong>
					<b>b</b>ingwa is a command-line tool used to carry out meta-analysis of genome-wide association studies.
					It is particularly suited for use with the program <a href="">SNPTEST</a>, but can also be used with other programs such as <a href="">MMM</a>.
					Key features of bingwa include:
				</p>
				<ul>
					<li>
						Computation of fixed-effect frequentist and flexible bayesian meta-analysis.
					</li>
					<li>
						Support 
					</li>
				</ul>
                <p>
                    <strong>Change history.</strong>
					<br />
					<em>Important!</em>
					This page documents version 2 of bingwa which is currently <b>experimental</b>.  This means we expect some features not to work, or
					not to work well, or to work wrongly, or to destroy your computer or sanity.
                </p>
                <p>
                    <strong>Acknowledgements.</strong> The following people contributed to the methods, design and implementation of this software:
                </p>
                <p>
                    <ul>
                        <li>
                            <a href="http://www.well.ox.ac.uk/~gav/">Gavin Band</a>
                        </li>
                        <li>
                            <a href="http://www.well.ox.ac.uk/chris-spencer">Chris Spencer</a>
                        </li>
                        <li>
                            <a href="http://www.well.ox.ac.uk/si-quang-le-2">Quang Si Le</a>
                        </li>
                        <li>
                            <a href="http://www.helsinki.fi/~mjxpirin/">Matti Pirinen</a>
                        </li>
                        <li>
                            <a href="http://www.well.ox.ac.uk/geraldine-clarke">Geraldine Clarke</a>
                        </li>
					</ul>
                </p>
                <p>
                    <strong>Contact.</strong> For more information or questions, please contact the oxstatgen mailing list at
                    <pre class="monospace">
                        oxstatgen (at) jiscmail.ac.uk
                    </pre>
                </p>
            </div>
            <div id="documentation" style="display:none">
                <div class="display_diagram_right">
                    <!--><object data="style/images/inthinnerator-flow.svg"  type="image/svg+xml" align="center">
					</object>-->
                </div>
                <p>
                    The general process followed by bingwa is shown on the right.  In brief, bingwa proceeds by
					<ol>
						<li>Loading association test data from a set of cohorts (e.g. SNPTEST files).</li>
						<li>Matching variants between cohorts.</li>
						<li>Computing frequentist meta-analysis and one or more Bayes factors.</li>
						<li>Writing results.</li>
					</ol>
                </p>
			</div>
            <div id="tutorial" style="display:none">
                <div class="tutorial_header">
                    This page shows common command lines to use inthinnerator.
					We assume the program is being run from a directory containing a file <span class="monospace">snps.txt</span>
					containing a list of SNPs to be picked from.  This file should look like this:
                    <div class="terminal_display">
<pre>$ head snps.txt
SNPID rsid chromosome position alleleA alleleB
snp1 kgp14987749 22 16152031 A C
snp2 rs150880246 22 16152288 A G
snp3 rs144530981 22 16152415 G A
snp4 rs142259989 22 16154873 T G
snp5 rs186282246 22 16155262 C T
snp6 22-16156144 22 16156144 GC G
snp7 22-16156442 22 16156442 G GCGT
snp8 22-16156727 22 16156727 T TG
snp9 22-16158548 22 16158548 ACT A</pre>
                    </div>
				</p>
				<p>
					Also, we assume the directory contains files <span class="monospace">genetic_map_chr1.txt</span>, <span class="monospace">genetic_map_chr2.txt</span>, ...
					containing a recombination map.  Suitable recombination map files can be downloaded
					from the <a href="https://mathgen.stats.ox.ac.uk/genetics_software/shapeit/shapeit.html#gmap">SHAPEIT website</a>
					or the <a href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#reference">IMPUTE website</a>.
				</p>
				<p>
					See also the <a href="inthinnerator_help.txt">list of options</a> and the <a href="#file_formats">file formats page</a>.
				</p>
                </div>
                <div class="tutorial_separator">
                </div>
                <div class="task">
                    <div class="task_name">
                        View the program usage page
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -help | less
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Thin SNPs based on physical distance
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -g snps.txt -min-distance 10kb -o thinned.out
                    </div>
                    <div class="task_notes">
                        <b>Note</b>: Inthinnerator understand distances expressed in base pairs ("bp"),
						kb, or Mb.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Thin SNPs based on recombination distance
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -g snps.txt -map genetic_map_chr#.txt -min-distance 0.125cM -o thinned.out
                    </div>
                    <div class="task_notes">
                        <b>Note</b>: Because the genetic map files don't contain chromosome information,
						you usually need to make inthinnerator infer the chromosome by using the # wildcard
						character, as shown above.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Thin SNPs based on recombination distance with a margin
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -g snps.txt -map genetic_map_chr#.txt -min-distance 0.125cM+25kb -o thinned.out
                    </div>
                    <div class="task_notes">
                        <b>Note</b>: The above command means that the interval around each picked SNP will extend out
						0.125cM in each direction.  The region is then expanded by 25kb in each direction.  This choice
						corresponds to the region definitions used in the <a href="http://www.well.ox.ac.uk/wtccc2/ms">IMSGC / WTCCC2 Multiple sclerosis GWAS paper</a>.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
						Write only the picked SNPs
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -g snps.txt -map genetic_map_chr#.txt -min-distance 0.125cM+25kb -o thinned.out -suppress-excluded
                    </div>
					<div class='task_notes'>
                        <b>Note</b>: Without the <span class="option">-suppress-excluded</span> flag, inthinnerator writes both the picked SNPs and the excluded ones.
					</div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Pick by rank
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -g snps.txt -map genetic_map_chr#.txt -min-distance 0.125cM+25kb -o thinned.out -rank &lt;filename&gt; -rank-column bf
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -g snps.txt -map genetic_map_chr#.txt -min-distance 0.125cM+25kb -o thinned.out -rank &lt;filename&gt; -rank-column -pvalue
                    </div>
                    <div class="task_notes">
						<b>Note</b>: the -rank option specifies which must have columns SNPID, rsid, chromosome, position, alleleA and alleleB,
						and the column named by the <span class="option">-rank-column</span> option.  Inthinnerator will pick the SNP with
						highest rank first, followed by the SNP with next highest rank among those still included, etc.  If the rank column
						name is preceded by a minus sign, ranks are negated before picking.  This can be useful for ranking by p-value.
                    </div>
                </div>
                <div class="task">
                    <div class="task_name">
                        Annotate the output with nearby genes.
                    </div>
                    <div class="task_command_line">
                        $ inthinnerator -g snps.txt -map genetic_map_chr#.txt -min-distance 0.125cM+25kb -o thinned.out -suppress-excluded
						-genes &lt;genes filename&gt;
                    </div>
                    <div class="task_notes">
						<b>Note</b>: Currently, the file passed to <span class="option">-genes</span> is expected to be in a similar format
						to the refGene.txt file downloadable from UCSC
						genome browser, but with a header line included.  Inthinnerator looks for and uses
						the chrom, txStart, txEnd and name2 columns of that file.
                    </div>
                </div>
            </div>
            <div id="file_formats" style="display:none">
                <div>
                    Inthinnerator shares a codebase with <a href="http://www.well.ox.ac.uk/~gav/qctool">qctool</a> and supports
					a similar set of file formats.  File types are detected based on the file extensions according to the following table.
                    <table class="file_formats">
                        <tr>
                            <th>
                                format
                            </th>
                            <th>
                                recognised extension(s)
                            </th>
                            <th>
                                notes
                            </th>
                        </tr>
                        <tr>
                            <td>
                                <a href="http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html">GEN</a>,
								<a href="http://www.well.ox.ac.uk/~gav/bgen_format/bgen_format.html">BGEN</a>,
								<a href="http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41">VCF</a>
								formats
                            </td>
                            <td>
                                <span class="monospace">.gen</span>,
								<span class="monospace">.gen.gz</span>,
								<span class="monospace">.bgen</span>,
								<span class="monospace">.vcf</span>,
								<span class="monospace">.vcf.gz</span>
                            </td>
                            <td>
								These can be supplied to the <span class="monospace">-g</span> option.  See the <a href="http://www.well.ox.ac.uk/~gav/qctool#file_formats">qctool webpage</a>
								for more detailed information on support for these formats.
                            </td>
                        </tr>
                        <tr>
                            <td>
                                Inthinnerator flat-file format
                            </td>
                            <td>
                                <span class="monospace">.txt</span>,
                                <span class="monospace">.csv</span>,
								<span class="monospace">.tsv</span>
                            </td>
                            <td>
								This is the most flexible format to supply to the <span class="monospace">-g</span>.  It consists of six columns,
								<span class="monospace">SNPID, rsid, chromosome, position, alleleA, alleleB</span>.  Optional additional columns include
									a <em>rank</em> column (specified by name using the <span class="monospace">-rank-column</span> option) and
									a <em>tag</em> column (specified using the <span class="monospace">-tag-column</span> option.)  Other columns
									are ignored.
									Inthinnerator expects space-, comma- or tab-separated file according to the file extension.
                            </td>
                        </tr>
                    </table>
                </div>
				<h3>Output file formats</h3>
				<div>
				<p>Bingwa can either output to a flat file or to a table in a sqlite database.  Both have pros and cons.</p>
				<table class="file_formats">
					<tr><th>Output type</th><th>Pros</th><th>Cons</th></tr>
					<tr>
						<td>Flat file</td>
						<td>Simple, easy-to use.  Splitting jobs into chunks allows for massive parallelism e.g. via a compute cluster.</td>
						<td>For large jobs, flat files can be awkward to analyse downstream because of large file sizes and potential
							split into chunks.</td>
					</tr>
					<tr>
						<td><a href="http://www.sqlite.org">Sqlite database</a></td>
						<td>Allows for flexible access to genome-wide results via the power of SQL.  Indices on position, rsid or other fields
							are used provide random access to scan results.</td>
						<td>Not suited to large parallel jobs (though some parallelism is supported).  Database should not be created on an NFS drive
							(but, once written, can be read across NFS).
						</td>
					</tr>
				</table>
			</div>
            </div>
            <div id="download" style="display:none">
                <div>
                    <p>
                        Inthinnerator is available either as binaries or as source code.
                    </p>
                    <h3>
                        Binaries
                    </h3>
                    <p>
                        Pre-compiled binaries are available for the following platforms.
                    </p>
                    <table>
                        <tr>
                            <th>
                                Version
                            </th>
                            <th>
                                Platform
                            </th>
                            <th>
                                File
                            </th>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                Linux <small>x86-64 static build</small>
                            </td>
                            <td>
                                <a href="resources/inthinnerator_v2.0-dev-linux-x86_64.tgz">inthinnerator_v2.0-dev-linux-x86_64.tgz</a>
                            </td>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                Linux <small>alternative build</small>
                            </td>
                            <td>
                                <a href="resources/inthinnerator_v2.0-dev-scientific-linux-x86_64.tgz">inthinnerator_v2.0-dev-scientific-linux-x86_64.tgz</a>
                            </td>
                        </tr>
                        <tr>
                            <td>v2-dev<sup>&dagger;</sup></td>
                            <td>
                                Mac OS X
                            </td>
                            <td>
                                <a href="resources/inthinnerator_v2.0-dev-osx.tgz">inthinnerator_v2.0-dev-osx.tgz</a>
                            </td>
                        </tr>
                    </table>
                    <p>
					<sup>&dagger;</sup><span class="smalltext">This version of inthinnerator is considered experimental.
					</p>
                    <p>
                        To run inthinnerator, download the relevant file and extract it as follows.
                    </p>
                    <pre>
$ tar -xzf inthinnerator_v2.0-dev-[machine].tgz
$ cd inthinnerator_v2.0-dev-[machine]
$ ./inthinnerator_v2.0-dev -help
</pre>
                    <h3>
                        Source
                    </h3>
                    <p>
                        The source code to inthinnerator is available as part of the <a href="http://bitbucket.org/gavinband/qctool">qctool package on bitbucket</a>.
						See <a href="http://www.well.ox.ac.uk/~gav/qctool#download">this page</a> for details.
					</p>
                </div>
            </div>
        </div>
    </body>
</html>
